control-plane/api-gateway/controllers/gateway_controller_integration_test.go

```diff
@@ -0,0 +1,671 @@
+)
+
+func TestControllerDoesNotInfinitelyReconcile(t *testing.T) {
+	s := runtime.NewScheme()
+	require.NoError(t, clientgoscheme.AddToScheme(s))
+	require.NoError(t, gwv1alpha2.Install(s))
+	require.NoError(t, gwv1beta1.Install(s))
+	require.NoError(t, v1alpha1.AddToScheme(s))
+
+	k8sClient := registerFieldIndexersForTest(fake.NewClientBuilder().WithScheme(s)).Build()
+	consulTestServerClient := test.TestServerWithMockConnMgrWatcher(t, nil)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	logger := logrtest.New(t)
+
+	cacheCfg := cache.Config{
+		ConsulClientConfig:  consulTestServerClient.Cfg,
+		ConsulServerConnMgr: consulTestServerClient.Watcher,
+		Logger:              logger,
+	}
+	resourceCache := cache.New(cacheCfg)
+
+	gwCache := cache.NewGatewayCache(ctx, cacheCfg)
+
+	gwCtrl := GatewayController{
+		HelmConfig:            common.HelmConfig{},
+		Log:                   logger,
+		Translator:            common.ResourceTranslator{},
+		cache:                 resourceCache,
+		gatewayCache:          gwCache,
+		Client:                k8sClient,
+		allowK8sNamespacesSet: mapset.NewSet(),
+		denyK8sNamespacesSet:  mapset.NewSet(),
+	}
+
+	go func() {
+		resourceCache.Run(ctx)
+	}()
+
+	resourceCache.WaitSynced(ctx)
+
+	gwSub := resourceCache.Subscribe(ctx, api.APIGateway, gwCtrl.transformConsulGateway)
+	httpRouteSub := resourceCache.Subscribe(ctx, api.HTTPRoute, gwCtrl.transformConsulHTTPRoute(ctx))
+	tcpRouteSub := resourceCache.Subscribe(ctx, api.TCPRoute, gwCtrl.transformConsulTCPRoute(ctx))
+	inlineCertSub := resourceCache.Subscribe(ctx, api.InlineCertificate, gwCtrl.transformConsulInlineCertificate(ctx))
+
+	k8sGWObj := createAllFieldsSetAPIGW(t, ctx, k8sClient)
+
+	// reconcile so we add the finalizer
+	_, err := gwCtrl.Reconcile(ctx, reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: k8sGWObj.Namespace,
+			Name:      k8sGWObj.Name,
+		},
+	})
+	require.NoError(t, err)
+
+	// reconcile again so that we get the creation with the finalizer
+	_, err = gwCtrl.Reconcile(ctx, reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: k8sGWObj.Namespace,
+			Name:      k8sGWObj.Name,
+		},
+	})
+	require.NoError(t, err)
+
+	cert := createAllFieldsSetCert(t, ctx, k8sClient)
+	httpRouteObj := createAllFieldsSetHTTPRoute(t, ctx, k8sClient, k8sGWObj)
+	tcpRouteObj := createAllFieldsSetTCPRoute(t, ctx, k8sClient, k8sGWObj)
+	// reconcile again so that we get the route bound to the gateway
+	_, err = gwCtrl.Reconcile(ctx, reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: k8sGWObj.Namespace,
+			Name:      k8sGWObj.Name,
+		},
+	})
+	require.NoError(t, err)
+
+	// reconcile again so that we get the route bound to the gateway
+	_, err = gwCtrl.Reconcile(ctx, reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: k8sGWObj.Namespace,
+			Name:      k8sGWObj.Name,
+		},
+	})
+	require.NoError(t, err)
+
+	wg := &sync.WaitGroup{}
+	wg.Add(3)
+	// wg.Add(4)
```

Comment:
for some reason the inline cert never sends an event down the channel. It looks like it's something to do with the fact that when it's initially created the gateway doesn't reference it, but once the gateway does reference it there's no changes made to the cert so it never actually gets a notification from consul: 
So we never enter the loop here:
https://github.com/hashicorp/consul-k8s/blob/NET-4231-halting-problem/control-plane/api-gateway/cache/consul.go#L250

because the translator function called never sees that the gateway references the secret:
https://github.com/hashicorp/consul-k8s/blob/NET-4231-halting-problem/control-plane/api-gateway/controllers/gateway_controller.go#L556

Reply:
talked about this on Slack, this is expected and due to the ordering of operations against Consul -- the cache notifier gets the cert notification and does the translation prior to writing the gateway back into the cache, so no reconciliation happens due to the cert creation, but the gateway update triggers it as soon as it references the cert